src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -1342,34 +1343,39 @@ void ZooKeeper::receiveEvent()
 
 void ZooKeeper::finalize(bool error_send, bool error_receive)
 {
-    {
-        std::lock_guard lock(push_request_mutex);
-
-        if (expired)
-            return;
-        expired = true;
-    }
+    if (expired)
```

Comment:
Possible race with line 1365 and line 1376 because of the absence of lock_guard ?